contracts/SinglePlayerCommit.sol

```diff
@@ -51,213 +61,470 @@ contract SinglePlayerCommit is Ownable {
+    function processCommitmentUser() public {
+        console.log("Processing commitment");
+        Commitment storage commitment = commitments[msg.sender];
 
-        if (met) {
-            penalty = 0;
-        } else {
-            penalty = stake;
-            // remove from committer's balance
-            _changeCommitterBalance(penalty, false);
-        }
+        commitment.met = commitment.reportedValue > commitment.goalValue;
 
-        emit CommitmentEnded(committer, met, penalty);
+        if (!commitment.met) {
+            _slashFunds(commitment.stake, msg.sender);
+        } 
+        
+        commitment.exists = false;
+        emit CommitmentEnded(msg.sender, commitment.met, commitment.stake);
     }
 
+    //TODO state change after transfer is not recommended
+    /// @notice Contract owner can withdraw funds not owned by committers. E.g. slashed from failed commitments
+    /// @param amount Amount of <token> to withdraw
+    /// @dev Check amount against slashedBalance, transfer amount and update slashedBalance
     function ownerWithdraw(uint256 amount) public onlyOwner returns (bool) {
-        uint256 available = token.balanceOf(address(this)).sub(committerBalance);
+        console.log("Received call for owner withdrawal for amount %s", amount);
 
-        require(amount <= available, "SPC::ownerWithdraw - not enough available balance");
+        require(amount <= slashedBalance, "SPC::ownerWithdraw - not enough available balance");
         require(token.transfer(msg.sender, amount), "SPC::ownerWithdraw - token transfer failed");
+        slashedBalance -= amount;
+
+        return true;
+    }
+
+    /// @notice Internal function to update balance of caller and total balance
+    /// @param amount Amount of <token> to deposit/withdraw
+    /// @param add Boolean toggle to deposit or withdraw
+    /// @dev Based on add param add or substract amount from msg.sender balance and total committerBalance
+    function _changeCommitterBalance(uint256 amount, bool add) internal returns (bool success) {
+        if (add) {
+            committerBalances[msg.sender] = committerBalances[msg.sender].add(amount);
+            totalCommitterBalance = totalCommitterBalance.add(amount);
+        } else {
+            committerBalances[msg.sender] = committerBalances[msg.sender].sub(amount);
+            totalCommitterBalance = totalCommitterBalance.sub(amount);
+        }
+
+        return true;
+    }
 
+    /// @notice Internal function to slash funds from user
+    /// @param amount Amount of <token> to slash
+    /// @param committer Address of committer
+    /// @dev Substract amount from committer balance and add to slashedBalance
+    function _slashFunds(uint256 amount, address committer) internal returns (bool success) {
+        require(committerBalances[committer] >= amount, "SPC::_slashFunds - funds not available");
+        _changeCommitterBalance(amount, false);
```

Comment:
Faulty logic. _changeCommitterBalance automatically slashes from msg.sender's balance. But should be based on committer address

Reply:
maybe simplest to revise `_changeCommitterBalance` to take an `address committer` argument?

Reply:
done